### DISCUSSION TITLE
Can Windows 10 Task View (Win-Tab) be enabled?

### DISCUSSION THREAD

**Discussion by lorzzn** [valinet/ExplorerPatcher#569] [General]
![img](https://s2.loli.net/2021/12/15/Mr21gXLivTctWBj.png)

As the picture and the question says

**Comment by valinet** [comment]
As this thread says: https://github.com/valinet/ExplorerPatcher/issues/56. A simple rule, ignored again.

I tested this a while ago. In `twinui.pcshell.dll`:

* There is a function called `IsUndockingAssetAvailable`. When that returns 1, a XAML (Windows 11) UI is created; when that returns 0, a DirectComposition (Windows 10) UI is created.
* In initial Windows 11 builds, there was a function called `GetMTVHostKind` which receieved a number x and called `IsUndockingAssetAvailable` for x = 1 and x = 4.
* Values of x correspond to: x = 1 = Alt-Tab view, x = 2 = Win-Tab view, x = 4 = Snap Assist
* So, `GetMTVHostKind` decided whether to create a Windows 10 or Windows 11 only for Alt-Tab view and Snap Assist view. For Win-Tab, `IsUnodckingAssetAvailable` is never called and a Windows 11 UI is always requested (as you have observed)
* What's worse, in recent builds, the `GetMTVHostKind` function has been inlined by the compiler and is now part of `CMultitaskingViewManager::_CreateMTVHost`.
* Take these instructions from `CMultitaskingViewManager::_CreateMTVHost`:
```
.text:000000018001E2E6                 call    ?EnsureXamlCoreForThreadIfNeeded@CMultitaskingViewManager@@AEAAXXZ ; CMultitaskingViewManager::EnsureXamlCoreForThreadIfNeeded(void)
.text:000000018001E2EB                 mov     ecx, edi
.text:000000018001E2ED                 sub     ecx, ebx
```
* When you invoke Win-Tab, after you return from `CMultitaskingViewManager::EnsureXamlCoreForThreadIfNeeded`, `edi` is copied to `ecx`, so `ecx` becomes a 2. If you change it to 1 (by modifying the registry when the breakpoint is hit in WinDbg), the if check will behave as if Alt-Tab was invoked when it comes to choosing a UI only. So, it would choose Windows 10 if you have a properly patched `IsUndockingAssetAvailable` (in EP, that is patched to return "Windows 10" for Alt-Tab when you use the Windows 10 window switcher). So, it should trick it into displaying a DirectComposition (Windows 10) UI for Win-Tab as well. (I also took care to patch the parameters for `CMultitaskingViewManager::_FindMTVHost` as well)
* Unfortunately, this crashes `explorer`, which is kind of expected. The check for `IsUndockingAssetAvailable` is skipped for case 2 (Win-Tab) in the code for good reason apparently: there is no DirectComposition (Windows 10) UI implementation for Win-Tab, apparently.
* So, the way it looks, it seems that they have removed the Windows 10 Win-Tab altogether. After the investigation, I wasn't able to have that come up, despite being successful in enabling the Windows 10 Alt-Tab and Snap Assist previously (and have exposed them as configurable settings in ExplorerPatcher). Or at least so it seems; anyway, it's pretty difficult anyway to develop a patching strategy for that, considering how volatile that function is and the fact that it would need modifications right in the middle of it.
* Beware that these things are largely kept in there so that Windows 11 Team (PPI) Edition works, which still features a Windows 10 UI mostly. Or at least that's my opinion after studying their binaries for quite a while now. Idk, it could be only speculation, I don't work at Microsoft, so I am just guessing.

So, tl;dr the Windows 10 Win-Tab seems to have been removed altogether. At least that's my conclusion at the moment after studying it for a bit. It kind of makes sense, as that featured the Timeline and whatnot, and those have been deprecated for a long time and are completely removed in Windows 11. In an ideal future, I plan on developing things in this area as well, but yeah... there's limited time available.

Also, please, next time, use the Discussion section for questions. Issues are for developers, confirmed bugs etc, not feature requests, questions, Q&As and so on. It's a simple rule, clearly states, why disregard that, I never understood...

**Comment by XuanChen1234** [comment]
You can ask from @Amrsatrio

**Comment by PandaFiredoge** [comment]
不行。
